Fix/segfault hamiltonian get hr

mikibonacci opened the pull request:
This fixes #439.

mikibonacci committed:
Allocating the arrays eigval_opt, eigval2 and utmp in the RAM.
i.e., after the variable declaration. This is needed to avoid
allocation on stack if the intel compiler is used, causing segmentation
fault for large systems. see issue https://github.com/wannier-developers/wannier90/issues/439.

mikibonacci committed:
This commit optimize the allocation of eigval_opt, utmp and eigval2, used in
the hamiltonian.F90 file. Basically the eigval_opt, utmp variable are
allocated only if have_disentangle is true.

mikibonacci committed:
adding dimension hint as a comment for eigval_opt, eigval2 and utmp
in hamiltonian.F90

mikibonacci committed:
making allocatable, in the routine sitesym_dis_extract_symmetry (sitesym.F90), the following variables:
umatnew, ZU, deltaU, carr. This is in direction of issue https://github.com/wannier-developers/wannier90/issues/439.

mikibonacci committed:
pre-commit fix

giovannipizzi reviewed (dismissed):
You also have to deallocate the variables

mikibonacci committed:
adding the deallocation of variables
- eigval2, eigval_opt, utmp in hamiltonian_get_hr
- umatnew, ZU, deltaU, carr in sitesym_dis_extract_symmetry

also, I moved the eigval2 allocation after the *if (ham_logical%have_ham_k) go to 100* line, because
if we go to 100, we do not need the eigval2 variale
